qutip/qip/qasm.py

```diff
@@ -0,0 +1,893 @@
+    qasm_lines = [line.strip() for line in qasm_lines]
+    qasm_lines = list(filter(lambda x: x[:2] != "//" and x != "", qasm_lines))
+
+    if version != "2.0":
+        raise NotImplementedError("QASM: Only OpenQASM 2.0 \
+                                  is currently supported.")
+
+    if qasm_lines.pop(0) != "OPENQASM 2.0;":
+        raise SyntaxError("QASM: File does not contain QASM 2.0 header")
+
+    qasm_obj = QasmProcessor(qasm_lines, mode=mode, version=version)
+    qasm_obj.commands = _tokenize(qasm_obj.commands)
+
+    qasm_obj._process_includes()
+
+    qasm_obj._initialize_pass()
+    qc = QubitCircuit(qasm_obj.num_qubits, num_cbits=qasm_obj.num_cbits)
+
+    qasm_obj._final_pass(qc)
+
+    return qc
+
+
+_GATE_NAME_TO_QASM_NAME = {
+    "QASMU": "U",
+    "RX": "rx",
+    "RY": "ry",
+    "RZ": "rz",
+    "SNOT": "h",
+    "X": "x",
+    "Y": "y",
+    "Z": "z",
+    "S": "s",
+    "T": "t",
+    "CRZ": "crz",
+    "CNOT": "cx",
+    "TOFFOLI": "ccx"
+}
+
+
+class QasmOutput():
+    """
+    Class for QASM export.
+
+    Parameters
+    ----------
+    version: str, optional
+        OpenQASM version, currently must be "2.0" necessarily.
+    """
+
+    def __init__(self, version="2.0"):
+        self.version = version
+        self.lines = []
+        self.gate_name_map = deepcopy(_GATE_NAME_TO_QASM_NAME)
+
+    def output(self, line="", n=0):
+        '''
+        Pipe QASM output string to QasmOutput's lines variable.
+
+        Parameters
+        ----------
+        line: str, optional
+            string to be appended to QASM output.
+        n: int, optional
+            number of blank lines to be appended to QASM output.
+        '''
+
+        if line:
+            self.lines.append(line)
+        self.lines = self.lines + [""] * n
+
+    def _flush(self):
+        '''
+        Resets QasmOutput variables.
+        '''
+
+        self.lines = []
+        self.gate_name_map = deepcopy(_GATE_NAME_TO_QASM_NAME)
+
+    def _qasm_output(self, qc):
+        '''
+        QASM output handler.
+
+        Parameters
+        ----------
+        qc: :class:`.QubitCircuit`
+            circuit object to produce QASM output for.
+        '''
+
+        self._flush()
+
+        self.output("// QASM 2.0 file generated by QuTiP", 1)
+
+        if self.version == "2.0":
+            self.output("OPENQASM 2.0;")
+        else:
+            raise NotImplementedError("QASM: Only OpenQASM 2.0 \
+                                      is currently supported.")
+
+        self.output('include "qelib1.inc";', 1)
+
+        qc._qasm_output(self)
+
+        return self.lines
+
+
+def print_qasm(qc):
+    '''
+    Print QASM output of circuit object.
+
+    Parameters
+    ----------
+    qc: :class:`.QubitCircuit`
+        circuit object to produce QASM output for.
+    '''
+
+    qasm_out = QasmOutput("2.0")
+    lines = qasm_out._qasm_output(qc)
+    for line in lines:
+        print(line)
+
+
+def str_qasm(qc):
```

Comment:
The name `str_qasm` is a little bit confusing to me. Probably `circuit_to_qasm_str`? Your take.

Reply:
Sounds good!